flocker/node/script.py

```diff
@@ -281,143 +319,282 @@ def get_service(self, reactor, options):
+
+    :ivar name: The human-meaningful name of this storage backend.
+    :ivar needs_reactor: A flag which indicates whether this backend's API
+        factory needs to have a reactor passed to it.
+    :ivar needs_id: A flag which indicates whether this backend's API factory
+        needs to have the cluster's unique identifier passed to it.
+    :ivar api_factory: An object which can be called with some simple
+        configuration data and which returns the API object implementing this
+        storage backend.
+    :ivar deployer_type: A constant from ``DeployerType`` indicating which kind
+        of ``IDeployer`` the API object returned by ``api_factory`` is usable
+        with.
+    """
+    name = field(type=unicode, mandatory=True)
+    needs_reactor = field(type=bool, mandatory=True)
+    # XXX Eventually everyone will take cluster_id so we will throw this flag
+    # out.
+    needs_id = field(type=bool, mandatory=True)
+    api_factory = field(mandatory=True)
+    deployer_type = field(mandatory=True)
+
+# These structures should be created dynamically to handle plug-ins
+_DEFAULT_BACKENDS = [
+    # P2PManifestationDeployer doesn't current know anything about
+    # cluster_uuid.  It probably should so that it can make sure it
+    # only talks to other nodes in the same cluster (maybe the
+    # authentication layer would mostly handle this but maybe not if
+    # you're slightly careless with credentials - also ZFS backend
+    # doesn't use TLS yet).
+    BackendDescription(
+        name=u"zfs", needs_reactor=True, needs_id=False,
+        api_factory=_zfs_storagepool, deployer_type=u"p2p",
+    ),
+    BackendDescription(
+        name=u"loopback", needs_reactor=False, needs_id=False,
+        api_factory=LoopbackBlockDeviceAPI.from_path, deployer_type=u"block",
+    ),
+
+    # BackendDescription(
+    #     name=u"openstack", needs_reactor=False, needs_id=True,
+    #     factory=cinder_from_configuration, deployer_type=u"block",
+    # ),
+]
+
+_DEFAULT_DEPLOYERS = {
+    'p2p': lambda api, **kw:
+        P2PManifestationDeployer(volume_service=api, **kw),
+    'block': lambda api, **kw:
+        BlockDeviceDeployer(block_device_api=api, **kw),
+}
 
-    :param dict dataset_configuration: Desired configuration for a dataset
-        deployer.
-    :param VolumeService dataset_configuration: An already started volume
-        service.
 
-    :return: A callable which can be called with a node UUID and hostname to
-        create a dataset deployer.
+class AgentService(PRecord):
     """
-    backend_to_deployer_factory = {
-        'zfs': zfs_dataset_deployer,
-        'loopback': loopback_dataset_deployer,
-    }
-    backend = dataset_configuration['backend']
-    deployer_factory = backend_to_deployer_factory[backend]
-    return deployer_factory(
-        volume_service=volume_service
+    :ivar PVector backends: ``BackendDescription`` instances describing how to
+        use each available storage backend.
```

Comment:
:question: There are quite a few missing `ivar` s missing here. Add them if you feel inclined.